merkle-patricia-tree/src/Data/Tree/MerklePatricia.hs

```diff
@@ -381,14 +418,110 @@ emptyRoot = hashNode "" []
 
 -- Walks a proof backwards and returns a root with element and root without element.
 --
--- The 'root with element' is calculated by walking the proofs from the leaf,
--- reconstructing the tree as the proof is walked backwards. This can be
--- compared to a root hash to prove existence.
---
--- The 'root without element' is calculated by walking the proofs from the leaf,
--- but, discarding the leaf itself. The resulting root can be compared to a
--- known root hash to see whether the proof is a direct extension of a given
--- MPT.
+-- Let's take an example, with the following MPT of size 5 and all values are
+-- simply units (see also 'pretty' below for explanation about how to read the
+-- pretty-printed MPTs). We'll walk the proof for '00100110', so let's also put
+-- side-by-side the same MPT with that element removed as it helps reasonning.
+--
+--     #dddda:
+--     #a7b9e: \0
+--     #93119:  \0100110 ↦ ()              #6d40d:
+--     #38e90:  \1000110 ↦ ()              #ceb6f: \01000110 ↦ ()
+--     #fa61b: \1                          #fa61b: \1
+--     #3e241:  \0                         #3e241:  \0
+--     #25f0c:   \000110 ↦ ()              #25f0c:   \000110 ↦ ()
+--     #d0bd1:   \100110 ↦ ()              #d0bd1:   \100110 ↦ ()
+--     #38e90:  \1000110 ↦ ()              #38e90:  \1000110 ↦ ()
+--
+-- Let's now consider the proof for the key '00100110':
+--
+-- - (ø, [(1,fa61b)])
+-- - (ø, [(1,38e90)])
```

Comment:
Why not use the normal `0` here?

Reply:
Because it's not a 0, it's an empty set! Since there's no prefix. Maybe not the best choice of symbol...